sunpy/net/scraper_utils.py

```diff
@@ -0,0 +1,150 @@
+import calendar
+from datetime import datetime, timedelta
+
+from dateutil.relativedelta import relativedelta
+
+from sunpy.extern.parse import parse
+from sunpy.time import TimeRange
+
+TIME_QUANTITIES = {'day': timedelta(days=1),
+                   'hour': timedelta(hours=1),
+                   'minute': timedelta(minutes=1),
+                   'second': timedelta(seconds=1),
+                   'millisecond': timedelta(milliseconds=1)}
+
+__all__ = ["extract_timestep", "date_floor", "check_timerange", "get_timerange_from_exdict"]
+
+def extract_timestep(directoryPattern):
+    """
+    Obtain the smaller time step for the given pattern.
+
+    Parameters
+    ----------
+    directoryPattern : `str`
+        The pattern containing the datetime-formats
+
+    Returns
+    -------
+    `dateutil.relativedelta.relativedelta`
+        The smallest timestep unit from the provided pattern.
+    """
+    if "%S" in directoryPattern:
+        return relativedelta(seconds=1)
+    elif "%M" in directoryPattern:
+        return relativedelta(minutes=1)
+    elif any(hour in directoryPattern for hour in ["%H"]):
+        return relativedelta(hours=1)
+    elif any(day in directoryPattern for day in ["%d", "%j"]):
+        return relativedelta(days=1)
+    elif any(month in directoryPattern for month in ["%b", "%B", "%m"]):
+        return relativedelta(months=1)
+    elif any(year in directoryPattern for year in ["%Y", "%y"]):
+        return relativedelta(years=1)
+    else:
+        return None
+
+def date_floor(date, timestep):
+    """
+    Return the "floor" of the given date and time step.
+
+    Parameters
+    ----------
+    datetime : `datetime.datetime` or `astropy.time.Time`
+        The date to floor
+    timestep : `dateutil.relativedelta.relativedelta`
+        The smallest time step to floor
+    Returns
+    -------
+    `datetime.datetime`
+        The time floored at the given time step
+    """
+    date_parts = [int(p) for p in date.strftime('%Y,%m,%d,%H,%M,%S').split(',')]
+    date_parts[-1] = date_parts[-1] % 60
+    date = datetime(*date_parts)
+    orig_time_tup = date.timetuple()
+    time_tup = [orig_time_tup.tm_year, orig_time_tup.tm_mon, orig_time_tup.tm_mday,
+                orig_time_tup.tm_hour, orig_time_tup.tm_min, orig_time_tup.tm_sec]
+    if timestep == relativedelta(minutes=1):
+        time_tup[-1] = 0
+    elif timestep == relativedelta(hours=1):
+        time_tup[-2:] = [0, 0]
+    elif timestep == relativedelta(days=1):
+        time_tup[-3:] = [0, 0, 0]
+    elif timestep == relativedelta(months=1):
+        time_tup[-4:] = [1, 0, 0, 0]
+    elif timestep == relativedelta(years=1):
+        time_tup[-5:] = [1, 1, 0, 0, 0]
```

Comment:
I think I'm originally responsible for this mess I think it can be done in cleaner way something like

```python
def date_floor(date, step):
    floor_date = date.copy()
    if step >= relativedelta(minutes=1):
        floor_date.replace(minutes=0)
    if step >= relativedelta(hours=1):
        floor_date.replace(hours=0)
    ...
    return floor_date
```

Reply:
This would be nice to fix.